src/ServiceControl.Monitoring/QueueLength/QueueLengthDataStore.cs

```diff
@@ -11,22 +13,59 @@ public QueueLengthDataStore(IQueueLengthCalculator calculator)
             this.calculator = calculator;
         }
 
-        public IEnumerable<KeyValuePair<string, JObject>> Current
+        public Dictionary<string, Dictionary<DateTime, double>> GetQueueLengths(DateTime now)
         {
-            get { return calculator.GetQueueLengths().Select(kvp => new KeyValuePair<string, JObject>(kvp.Key, new JObject(new JProperty("Count", kvp.Value)))); }
+            var results = new Dictionary<string, Dictionary<DateTime, double>>();
+
+            foreach (var queueLength in queueLengths)
+            {
+                var endpointName = queueLength.Key;
+
+                ConcurrentDictionary<DateTime, double> endpointData;
+
+                if (queueLengths.TryGetValue(endpointName, out endpointData))
+                {
+                    var endpointDataSnapshot = endpointData.ToArray();
+
+                    var points = endpointDataSnapshot
+                        .Where(kvp => kvp.Key > now.Subtract(HistoricalPeriod))
+                        .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+                    results.Add(endpointName, points);
+
+                    var outdatedPoints = endpointData.Where(kvp => kvp.Key <= now.Subtract(HistoricalPeriod));
+
+                    foreach (var outdatedPoint in outdatedPoints)
+                    {
+                        double removedValue;
+                        endpointData.TryRemove(outdatedPoint.Key, out removedValue);
+                    }
+                }
+            }
+
+            return results;
         }
 
-        public void Store(JObject data)
+        public void Store(string endpointName, JObject data)
         {
-            var counters = (JArray) data["Counters"] ?? EmptyArray;
-            UpdateSends(counters);
+            var updatedEndpoints = new List<string>();
 
-            var gauges = (JArray) data["Gauges"] ?? EmptyArray;
-            UpdateReceives(gauges);
+            var counters = (JArray) data["Counters"] ?? new JArray();
```

Comment:
I don't like that this class knows about the message format. This is all going to quietly fail if we update the other end.